src/framework/global/concurrency/taskscheduler.h

```diff
@@ -0,0 +1,211 @@
+        {
+            const std::lock_guard lock(m_mutex);
+            m_taskQueue.push(taskFunctor);
+        }
+        m_newTaskAvailableCv.notify_one();
+    }
+
+    template<typename FuncT, typename ... ArgsT, typename ReturnT = std::invoke_result_t<std::decay_t<FuncT>, std::decay_t<ArgsT>...> >
+    std::future<ReturnT> submit(FuncT&& task, ArgsT&&... args)
+    {
+        std::function<ReturnT()> taskFunctor = std::bind(std::forward<FuncT>(task), std::forward<ArgsT>(args)...);
+        std::shared_ptr<std::promise<ReturnT> > promise = std::make_shared<std::promise<ReturnT> >();
+        push([taskFunctor, promise] {
+            try {
+                if constexpr (std::is_void_v<ReturnT>) {
+                    std::invoke(taskFunctor);
+                    promise->set_value();
+                } else {
+                    promise->set_value(std::invoke(taskFunctor));
+                }
+            } catch (...) {
+                try {
+                    promise->set_exception(std::current_exception());
+                } catch (...) {
+                    LOGE() << "Unable to schedule a task";
+                }
+            }
+        });
+
+        return promise->get_future();
+    }
+
+    void waitForAllTasksComplete()
+    {
+        m_isWaitingForAllTasksDone = true;
+        std::unique_lock<std::mutex> tasks_lock(m_mutex);
+        m_taskFinishedCv.wait(tasks_lock, [this] { return m_taskQueue.empty(); });
+        m_isWaitingForAllTasksDone = false;
+    }
+
+    const std::set<std::thread::id>& threadIdSet() const
+    {
+        static std::set<std::thread::id> result;
+
+        if (result.empty()) {
+            for (thread_pool_size_t i = 0; i < m_threadPoolSize; ++i) {
+                result.insert(m_threadPool[i].get_id());
+            }
+        }
+
+        return result;
+    }
+
+    bool containsThread(const std::thread::id& id) const
+    {
+        const auto& idSet = threadIdSet();
+        return idSet.find(id) != idSet.cend();
+    }
+
+private:
+    void setupThreads()
+    {
+        m_isActive = true;
+        for (thread_pool_size_t i = 0; i < m_threadPoolSize; ++i) {
+            m_threadPool[i] = std::thread(&TaskScheduler::th_workerLoop, this);
+        }
+    }
+
+    void terminateThreads()
+    {
+        m_isActive = false;
+        m_newTaskAvailableCv.notify_all();
+        for (thread_pool_size_t i = 0; i < m_threadPoolSize; ++i) {
+            m_threadPool[i].join();
+        }
+    }
+
+    thread_pool_size_t vaildateThreadPoolCapacity(const thread_pool_size_t desiredThreadCount)
+    {
+        thread_pool_size_t maxCapacity = std::thread::hardware_concurrency();
+
+        if (maxCapacity <= 1) {
+            return 1;
+        }
+
+        thread_pool_size_t optimalCapacity = maxCapacity / 2;
+
+        if (desiredThreadCount <= 0) {
+            return optimalCapacity;
+        }
+
+        return desiredThreadCount;

```

Comment:
This is *way* outside my area of expertise, but should there be a check like this here?
```c++
if (desiredThreadCount > optimalCapacity) {
    return optimalCapacity;
}
```
Or perhaps this?
```c++
if (desiredThreadCount > maxCapacity) {
    return maxCapacity;
}
```

Reply:
nope, number of allocated threads not necessarily should be <= of logical/physical cores. It might be deliberate choice of a consumer